### ISSUE TITLE
Failed to restore etcd snapshot -  mv: inter-device move failed: '/opt/rke/etcd-snapshots-restore/member' to '/var/lib/rancher/etcd/member'; unable to remove target: Directory not empty

### ISSUE THREAD

**Issue by cruikshj** [rancher/rke#2502]
**RKE version:**

 `v1.0.14`

**Docker version:**

```
Client: Docker Engine - Community
 Version:           20.10.3
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        48d30b5
 Built:             Fri Jan 29 14:33:21 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.3
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       46229ca
  Built:            Fri Jan 29 14:31:32 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)

Server:
 Containers: 17
  Running: 7
  Paused: 0
  Stopped: 10
 Images: 24
 Server Version: 20.10.3
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-70-generic
 Operating System: Ubuntu 20.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 5.806GiB
 Name: cyrodiil
 ID: J2RJ:3CC2:X7I3:HVRP:PIDB:KYKV:W2KA:YGYP:RXIG:Y72J:KJLS:VKAU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

```

**Operating system and kernel:**

```
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

`5.4.0-70-generic`

**Type/provider of hosts:**

`Hyper-V`

**cluster.yml file:**

```
nodes:
- address: 192.168.1.10
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - worker
  - etcd
  hostname_override: "cyrodiil"
  user: rke
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: ~/.ssh/id_rsa
  ssh_cert: ""
  ssh_cert_path: ""
  labels: {}
  taints: []
- address: 192.168.1.11
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - worker
  - etcd
  hostname_override: "rhodes"
  user: rke
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: ~/.ssh/id_rsa
  ssh_cert: ""
  ssh_cert_path: ""
  labels: {}
  taints: []
- address: 192.168.1.12
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - worker
  - etcd
  hostname_override: "roshar"
  user: rke
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: ~/.ssh/id_rsa
  ssh_cert: ""
  ssh_cert_path: ""
  labels: {}
  taints: []
services:
  etcd:
    snapshot: true
    retention: 24h
    creation: 6h
network:
  plugin: canal
ignore_docker_version: true
ingress:
  provider: nginx
  options: 
    use-forwarded-headers: "true"
cluster_name: "rancher"
```

**Steps to Reproduce:**

```
rke etcd snapshot-restore --config cluster.yml --name 2021-03-29T06:19:55Z_etcd
```

**Results:**

```
INFO[0000] Running RKE version: v1.0.14
INFO[0000] Restoring etcd snapshot 2021-03-29T06:19:55Z_etcd
INFO[0000] Successfully Deployed state file at [./cluster.rkestate]
INFO[0000] [dialer] Setup tunnel for host [192.168.1.12]
INFO[0000] [dialer] Setup tunnel for host [192.168.1.10]
INFO[0000] [dialer] Setup tunnel for host [192.168.1.11]
INFO[0001] Checking if container [cert-deployer] is running on host [192.168.1.12], try #1
INFO[0001] Checking if container [cert-deployer] is running on host [192.168.1.10], try #1
INFO[0001] Checking if container [cert-deployer] is running on host [192.168.1.11], try #1
INFO[0001] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.12]
INFO[0001] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.10]
INFO[0001] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.11]
INFO[0001] Starting container [cert-deployer] on host [192.168.1.12], try #1
INFO[0001] Starting container [cert-deployer] on host [192.168.1.11], try #1
INFO[0002] Starting container [cert-deployer] on host [192.168.1.10], try #1
INFO[0002] Checking if container [cert-deployer] is running on host [192.168.1.12], try #1
INFO[0002] Checking if container [cert-deployer] is running on host [192.168.1.10], try #1
INFO[0002] Checking if container [cert-deployer] is running on host [192.168.1.11], try #1
INFO[0007] Checking if container [cert-deployer] is running on host [192.168.1.12], try #1
INFO[0007] Removing container [cert-deployer] on host [192.168.1.12], try #1
INFO[0007] Checking if container [cert-deployer] is running on host [192.168.1.10], try #1
INFO[0007] Removing container [cert-deployer] on host [192.168.1.10], try #1
INFO[0007] Checking if container [cert-deployer] is running on host [192.168.1.11], try #1
INFO[0007] Removing container [cert-deployer] on host [192.168.1.11], try #1
INFO[0007] Stopping container [etcd] on host [192.168.1.10] with stopTimeoutDuration [5s], try #1
WARN[0007] Can't stop Docker container [etcd] for host [192.168.1.10]: Error response from daemon: No such container: etcd
INFO[0007] Stopping container [etcd] on host [192.168.1.10] with stopTimeoutDuration [5s], try #2
WARN[0007] Can't stop Docker container [etcd] for host [192.168.1.10]: Error response from daemon: No such container: etcd
INFO[0007] Stopping container [etcd] on host [192.168.1.10] with stopTimeoutDuration [5s], try #3
WARN[0007] Can't stop Docker container [etcd] for host [192.168.1.10]: Error response from daemon: No such container: etcd
INFO[0007] [etcd] starting backup server on host [192.168.1.10]
INFO[0007] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.10]
INFO[0007] Starting container [etcd-Serve-backup] on host [192.168.1.10], try #1
INFO[0008] [etcd] Successfully started [etcd-Serve-backup] container on host [192.168.1.10]
INFO[0013] [etcd] Get snapshot [2021-03-29T06:19:55Z_etcd] on host [192.168.1.11]
INFO[0013] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.11]
INFO[0013] Starting container [etcd-download-backup] on host [192.168.1.11], try #1
INFO[0013] [etcd] Successfully started [etcd-download-backup] container on host [192.168.1.11]
INFO[0013] Waiting for [etcd-download-backup] container to exit on host [192.168.1.11]
INFO[0013] Container [etcd-download-backup] is still running on host [192.168.1.11]
INFO[0014] Waiting for [etcd-download-backup] container to exit on host [192.168.1.11]
INFO[0014] Removing container [etcd-download-backup] on host [192.168.1.11], try #1
INFO[0014] [etcd] Get snapshot [2021-03-29T06:19:55Z_etcd] on host [192.168.1.12]
INFO[0014] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.12]
INFO[0014] Starting container [etcd-download-backup] on host [192.168.1.12], try #1
INFO[0014] [etcd] Successfully started [etcd-download-backup] container on host [192.168.1.12]
INFO[0014] Waiting for [etcd-download-backup] container to exit on host [192.168.1.12]
INFO[0014] Container [etcd-download-backup] is still running on host [192.168.1.12]
INFO[0015] Waiting for [etcd-download-backup] container to exit on host [192.168.1.12]
INFO[0015] Removing container [etcd-download-backup] on host [192.168.1.12], try #1
INFO[0015] Removing container [etcd-Serve-backup] on host [192.168.1.10], try #1
INFO[0015] [remove/etcd-Serve-backup] Successfully removed container on host [192.168.1.10]
INFO[0015] [etcd] Checking if all snapshots are identical
INFO[0015] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.10]
INFO[0016] Starting container [etcd-checksum-checker] on host [192.168.1.10], try #1
INFO[0016] [etcd] Successfully started [etcd-checksum-checker] container on host [192.168.1.10]
INFO[0016] Waiting for [etcd-checksum-checker] container to exit on host [192.168.1.10]
INFO[0016] Container [etcd-checksum-checker] is still running on host [192.168.1.10]
INFO[0017] Waiting for [etcd-checksum-checker] container to exit on host [192.168.1.10]
INFO[0017] Removing container [etcd-checksum-checker] on host [192.168.1.10], try #1
INFO[0017] [etcd] Checksum of etcd snapshot on host [192.168.1.10] is [b2f8a9709dbcc718bd1557cea3696cb2]
INFO[0017] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.11]
INFO[0017] Starting container [etcd-checksum-checker] on host [192.168.1.11], try #1
INFO[0018] [etcd] Successfully started [etcd-checksum-checker] container on host [192.168.1.11]
INFO[0018] Waiting for [etcd-checksum-checker] container to exit on host [192.168.1.11]
INFO[0018] Container [etcd-checksum-checker] is still running on host [192.168.1.11]
INFO[0019] Waiting for [etcd-checksum-checker] container to exit on host [192.168.1.11]
INFO[0019] Removing container [etcd-checksum-checker] on host [192.168.1.11], try #1
INFO[0019] [etcd] Checksum of etcd snapshot on host [192.168.1.11] is [b2f8a9709dbcc718bd1557cea3696cb2]
INFO[0019] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.12]
INFO[0019] Starting container [etcd-checksum-checker] on host [192.168.1.12], try #1
INFO[0020] [etcd] Successfully started [etcd-checksum-checker] container on host [192.168.1.12]
INFO[0020] Waiting for [etcd-checksum-checker] container to exit on host [192.168.1.12]
INFO[0020] Container [etcd-checksum-checker] is still running on host [192.168.1.12]
INFO[0021] Waiting for [etcd-checksum-checker] container to exit on host [192.168.1.12]
INFO[0021] Removing container [etcd-checksum-checker] on host [192.168.1.12], try #1
INFO[0021] [etcd] Checksum of etcd snapshot on host [192.168.1.12] is [b2f8a9709dbcc718bd1557cea3696cb2]
INFO[0021] Cleaning old kubernetes cluster
INFO[0021] [worker] Tearing down Worker Plane..
INFO[0021] [worker] Successfully tore down Worker Plane..
INFO[0021] [controlplane] Tearing down the Controller Plane..
INFO[0021] [controlplane] Successfully tore down Controller Plane..
INFO[0021] [etcd] Tearing down etcd plane..
INFO[0021] [etcd] Successfully tore down etcd plane..
INFO[0021] [hosts] Cleaning up host [192.168.1.11]
INFO[0021] [hosts] Cleaning up host [192.168.1.11]
INFO[0021] [hosts] Cleaning up host [192.168.1.10]
INFO[0021] [hosts] Cleaning up host [192.168.1.10]
INFO[0021] [hosts] Running cleaner container on host [192.168.1.10]
INFO[0021] [hosts] Running cleaner container on host [192.168.1.11]
INFO[0021] [hosts] Cleaning up host [192.168.1.12]
INFO[0021] [hosts] Cleaning up host [192.168.1.12]
INFO[0021] [hosts] Running cleaner container on host [192.168.1.12]
INFO[0021] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.12]
INFO[0021] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.10]
INFO[0021] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.11]
INFO[0021] Starting container [kube-cleaner] on host [192.168.1.12], try #1
INFO[0021] Starting container [kube-cleaner] on host [192.168.1.10], try #1
INFO[0021] Starting container [kube-cleaner] on host [192.168.1.11], try #1
INFO[0021] [kube-cleaner] Successfully started [kube-cleaner] container on host [192.168.1.12]
INFO[0021] Waiting for [kube-cleaner] container to exit on host [192.168.1.12]
INFO[0021] Container [kube-cleaner] is still running on host [192.168.1.12]
INFO[0021] [kube-cleaner] Successfully started [kube-cleaner] container on host [192.168.1.10]
INFO[0021] Waiting for [kube-cleaner] container to exit on host [192.168.1.10]
INFO[0021] Container [kube-cleaner] is still running on host [192.168.1.10]
INFO[0022] [kube-cleaner] Successfully started [kube-cleaner] container on host [192.168.1.11]
INFO[0022] Waiting for [kube-cleaner] container to exit on host [192.168.1.11]
INFO[0022] Container [kube-cleaner] is still running on host [192.168.1.11]
INFO[0022] Waiting for [kube-cleaner] container to exit on host [192.168.1.12]
INFO[0022] [hosts] Removing cleaner container on host [192.168.1.12]
INFO[0022] Removing container [kube-cleaner] on host [192.168.1.12], try #1
INFO[0022] [hosts] Removing dead container logs on host [192.168.1.12]
INFO[0022] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.12]
INFO[0022] Starting container [rke-log-cleaner] on host [192.168.1.12], try #1
INFO[0022] Waiting for [kube-cleaner] container to exit on host [192.168.1.10]
INFO[0022] [hosts] Removing cleaner container on host [192.168.1.10]
INFO[0022] Removing container [kube-cleaner] on host [192.168.1.10], try #1
INFO[0022] [hosts] Removing dead container logs on host [192.168.1.10]
INFO[0022] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.10]
INFO[0022] Starting container [rke-log-cleaner] on host [192.168.1.10], try #1
INFO[0023] [cleanup] Successfully started [rke-log-cleaner] container on host [192.168.1.12]
INFO[0023] Removing container [rke-log-cleaner] on host [192.168.1.12], try #1
INFO[0023] [remove/rke-log-cleaner] Successfully removed container on host [192.168.1.12]
INFO[0023] [hosts] Successfully cleaned up host [192.168.1.12]
INFO[0023] [cleanup] Successfully started [rke-log-cleaner] container on host [192.168.1.10]
INFO[0023] Removing container [rke-log-cleaner] on host [192.168.1.10], try #1
INFO[0023] Waiting for [kube-cleaner] container to exit on host [192.168.1.11]
INFO[0023] [hosts] Removing cleaner container on host [192.168.1.11]
INFO[0023] Removing container [kube-cleaner] on host [192.168.1.11], try #1
INFO[0023] [hosts] Removing dead container logs on host [192.168.1.11]
INFO[0023] [remove/rke-log-cleaner] Successfully removed container on host [192.168.1.10]
INFO[0023] [hosts] Successfully cleaned up host [192.168.1.10]
INFO[0023] Image [rancher/rke-tools:v0.1.66] exists on host [192.168.1.11]
INFO[0023] Starting container [rke-log-cleaner] on host [192.168.1.11], try #1
INFO[0024] [cleanup] Successfully started [rke-log-cleaner] container on host [192.168.1.11]
INFO[0024] Removing container [rke-log-cleaner] on host [192.168.1.11], try #1
INFO[0024] [remove/rke-log-cleaner] Successfully removed container on host [192.168.1.11]
INFO[0024] [hosts] Successfully cleaned up host [192.168.1.11]
INFO[0024] [etcd] Restoring [2021-03-29T06:19:55Z_etcd] snapshot on etcd host [192.168.1.10]
INFO[0024] Image [rancher/coreos-etcd:v3.4.3-rancher1] exists on host [192.168.1.10]
INFO[0024] Starting container [etcd-restore] on host [192.168.1.10], try #1
INFO[0024] [etcd] Successfully started [etcd-restore] container on host [192.168.1.10]
INFO[0024] Waiting for [etcd-restore] container to exit on host [192.168.1.10]
INFO[0024] Container [etcd-restore] is still running on host [192.168.1.10]
INFO[0025] Waiting for [etcd-restore] container to exit on host [192.168.1.10]
INFO[0025] Removing container [etcd-restore] on host [192.168.1.10], try #1
INFO[0025] [etcd] Removing snapshot [2021-03-29T06:19:55Z_etcd] from host [192.168.1.10]
INFO[0025] Image [rancher/coreos-etcd:v3.4.3-rancher1] exists on host [192.168.1.10]
INFO[0026] Starting container [etcd-remove-snapshot] on host [192.168.1.10], try #1
INFO[0026] [etcd] Successfully started [etcd-remove-snapshot] container on host [192.168.1.10]
INFO[0026] Waiting for [etcd-remove-snapshot] container to exit on host [192.168.1.10]
INFO[0026] Container [etcd-remove-snapshot] is still running on host [192.168.1.10]
INFO[0027] Waiting for [etcd-remove-snapshot] container to exit on host [192.168.1.10]
INFO[0027] Removing container [etcd-remove-snapshot] on host [192.168.1.10], try #1
INFO[0027] [etcd] Restoring [2021-03-29T06:19:55Z_etcd] snapshot on etcd host [192.168.1.11]
INFO[0027] Image [rancher/coreos-etcd:v3.4.3-rancher1] exists on host [192.168.1.11]
INFO[0027] Starting container [etcd-restore] on host [192.168.1.11], try #1
INFO[0027] [etcd] Successfully started [etcd-restore] container on host [192.168.1.11]
INFO[0027] Waiting for [etcd-restore] container to exit on host [192.168.1.11]
INFO[0027] Container [etcd-restore] is still running on host [192.168.1.11]
INFO[0028] Waiting for [etcd-restore] container to exit on host [192.168.1.11]
INFO[0028] Removing container [etcd-restore] on host [192.168.1.11], try #1
FATA[0028] [etcd] Failed to restore etcd snapshot: Failed to run etcd restore container, exit status is: 1, container logs: {"level":"info","ts":1617151493.177632,"caller":"membership/cluster.go:392","msg":"added member","cluster-id":"aa142edfdd923580","local-member-id":"0","added-peer-id":"218d8bfb33a29c6","added-peer-peer-urls":["https://192.168.1.12:2380"]}
{"level":"info","ts":1617151493.178334,"caller":"membership/cluster.go:392","msg":"added member","cluster-id":"aa142edfdd923580","local-member-id":"0","added-peer-id":"85b27104f8aed9bc","added-peer-peer-urls":["https://192.168.1.10:2380"]}
{"level":"info","ts":1617151493.178945,"caller":"membership/cluster.go:392","msg":"added member","cluster-id":"aa142edfdd923580","local-member-id":"0","added-peer-id":"ffbbc02c920c1179","added-peer-peer-urls":["https://192.168.1.11:2380"]}
{"level":"info","ts":1617151493.1935408,"caller":"snapshot/v3_snapshot.go:300","msg":"restored snapshot","path":"/opt/rke/etcd-snapshots/2021-03-29T06:19:55Z_etcd","wal-dir":"/opt/rke/etcd-snapshots-restore/member/wal","data-dir":"/opt/rke/etcd-snapshots-restore/","snap-dir":"/opt/rke/etcd-snapshots-restore/member/snap"}
mv: inter-device move failed: '/opt/rke/etcd-snapshots-restore/member' to '/var/lib/rancher/etcd/member'; unable to remove target: Directory not empty
```

**Notes:**

1. A power loss event resulted in one of the 3 nodes in this HA upstream RKE cluster to become corrupted. 
2. I attempted to follow this guide to remove and re-add the corrupt node: https://support.rancher.com/hc/en-us/articles/360055096171-How-to-remove-and-replace-an-unresponsive-control-plane-etcd-node-in-the-local-Rancher-server-cluster-provisioned-by-the-Rancher-Kubernetes-Engine-RKE-CLI#task-0-0.  However, the node was unable to join the cluster, and the rest of the cluster was not starting up.
3. I then attempted a snapshot restore using the command above and got the error.  
4. I then cleaned the bad node using https://github.com/rancherlabs/support-tools/tree/master/extended-rancher-2-cleanup and attempted a restore again and got the above error.
5. At this point, the failed restore wiped all three nodes and so I cleaned all three nodes using the above script and attempted another restore and got the same error.
6. I attempted to `rke up` a new cluster on these "cleaned" nodes.  The new cluster successfully started.
7. I then ran the restore against the new cluster, and got the same error.
8. I then attempted to download the latest `rke`, and did the same operations, but got the same error.
9. The files specified in the error do not exist on any node, nor the management machine, before or after the restore operation runs and fails.

I was unable to find much of anything regarding this error.  Any help is appreciated

**Comment by cruikshj** [comment]
I would go ahead and simply rebuild the upstream Rancher cluster from scratch, but according to this https://github.com/rancher/rancher/issues/16471, it seems as that I will also have to rebuild the downstream clusters, which is non trivial.

**Comment by superseb** [comment]
The only way to debug this would be to loop over `ls -la` over both the host directories and the mounted directories in the container to monitor the contents and see what is happening at what point and when the error occurs. If the error is `Directory not empty` while it is constantly, I would check the mount points on the host (possible different filesystems/devices?). In that case maybe you can share output of `df` and `mount`.

**Comment by cruikshj** [comment]
@superseb, thank you for taking a look.  I have been unable to see what is happening during the process.  It seems to fail pretty fast and then clean up the files as far as I can tell.

**Comment by cruikshj** [comment]
Today, I was able to narrow the problem down to just one of the 3 nodes by running the process with various combinations until 2/3 nodes worked.  I was unable to determine root cause of why it failed this way on the one node.  I ended up rebuilding the VM from scratch and adding it to the cluster without issue.  Unfortunately, the RKE error did not indicate the problem on a specific node, so I didn't immediately try to investigate using this approach.

Due to the restore, I did seem to lose my CA certificates for the Rancher endpoints, which causes all the downstream clusters to fail with a CA_CHECKSUM error.  I am still investigating this issue, but I am treating it as a distinct issue, so I will close this one.
